src/Development/IDE/Core/Rules.hs

```diff
@@ -398,6 +444,88 @@ loadGhcSession = do
         opts <- getIdeOptions
         return ("" <$ optShakeFiles opts, ([], Just val))
 
+getHiFileRule :: Rules ()
+getHiFileRule = defineEarlyCutoff $ \GetHiFile f -> do
+  session <- hscEnv <$> use_ GhcSession f
+  -- get all dependencies interface files, to check for freshness
+  (deps,_)<- use_ GetLocatedImports f
+  depHis  <- traverse (use GetHiFile) (mapMaybe (fmap artifactFilePath . snd) deps)
+
+  -- TODO find the hi file without relying on the parsed module
+  --      it should be possible to construct a ModSummary parsing just the imports
+  --      (see HeaderInfo in the GHC package)
+  pm      <- use_ GetParsedModule f
+  let hiFile = toNormalizedFilePath $
+            case ms_hsc_src ms of
+                HsBootFile -> addBootSuffix (ml_hi_file $ ms_location ms)
+                _ -> ml_hi_file $ ms_location ms
+      ms     = pm_mod_summary pm
+
+  case sequence depHis of
+    Nothing -> do
+          let d = mkInterfaceFilesGenerationDiag f "Missing interface file dependencies"
+          pure (Nothing, (d, Nothing))
+    Just deps -> do
+      gotHiFile <- getFileExists hiFile
+      if not gotHiFile
+        then do
+          let d = mkInterfaceFilesGenerationDiag f "Missing interface file"
+          pure (Nothing, (d, Nothing))
+        else do
+          hiVersion  <- use_ GetModificationTime hiFile
+          modVersion <- use_ GetModificationTime f
+          let sourceModified = LT == comparing modificationTime hiVersion modVersion
+          if sourceModified
+            then do
+              let d = mkInterfaceFilesGenerationDiag f "Stale interface file"
+              pure (Nothing, (d, Nothing))
+            else do
+              r <- liftIO $ loadInterface session ms deps
+              case r of
+                Right iface -> do
+                  let result = HiFileResult ms iface
+                  return (Just (fingerprintToBS (mi_mod_hash iface)), ([], Just result))
+                Left err -> do
+                  let diag = ideErrorWithSource (Just "interface file loading") (Just DsError) f . T.pack $ err
+                  return (Nothing, (pure diag, Nothing))
+
+mkInterfaceFilesGenerationDiag :: a -> String -> [(a, ShowDiagnostic, Diagnostic)]
+mkInterfaceFilesGenerationDiag f intro = mkDiag $ intro <> msg
+  where
+      msg = ": additional resource use while generating interface files in the background."
+      mkDiag = pure . ideErrorWithSource (Just "interface file loading") (Just DsInfo) f . T.pack
```

Comment:
Is this still going to produce the warnings that we discussed?

Reply:
Updated as suggested here - https://github.com/pepeiborra/ghcide/commit/9f1ce515140da365ba3023d9b68c8413928dda44#r37339599

Reply:
I've moved the interface loading diagnostics behind the recently added `--test` flag